refactor(Cash flow report)

lomamech opened the pull request:
- Implement Cashflow Report
- feat(bhCashboxSelect) implement the componant Cashbox Select
- Implement the test E2E for Cashflow Report

closes #1805

lomamech committed:
test(report cashflow)
Integrate test of regression tests based on the development data set

lomamech committed:
perf(CashFlowReport)
- Use AccountsExtra for to get the Opening Balance of an Account from a
  Date

- Improve the calculation of incomes cashflow report

lomamech committed:
test(cash flow report)
- Implement test of Cashflow report with expected of Data

lomamech committed:
fix(Cash flow report Regression tests)
Remove the property only for regression tests

lomamech committed:
test(Cash flow regression test): Update and reset data for cash flow
regression test

lomamech committed:
perf(Cash flow report): set the BHIMA similar style

jniles commented on client/src/js/components/bhCashboxSelect.js:
```diff
@@ -0,0 +1,55 @@
+        });
+        
+        $ctrl.cashboxes = cashboxes;
+      });
```
This should probably have a `.catch()` statement.

jniles commented on client/src/modules/templates/bhCashboxSelect.tmpl.html:
```diff
@@ -0,0 +1,21 @@
+    <ui-select name="cashbox_id" 
+      ng-model="$ctrl.cashboxId"
+      on-select="$ctrl.onSelect($item, $model)" required>
+      <ui-select-match placeholder="{{ 'FORM.SELECT.CASHBOX' | translate }}">{{$select.selected.hrlabel}}</span></ui-select-match>
```
This looks like you have a broken `</span>`.  You should remove it.

jniles commented on server/controllers/finance/reports/cashflow/index.js:
```diff
@@ -66,11 +69,12 @@ function processingCashflowReport(params) {
     .then((balance) => {
-      if (!balance.length) { balance[0] = { balance : 0, account_id : params.account_id }; }
-      glb.openningBalance = balance[0];
+      if (!balance) { balance = { balance : 0, account_id : params.account_id }; }
```
`AccountsExtra.getOpeningBalanceForDate()` will return an object.  See [this line](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/controllers/finance/accounts/extra.js#L152).  That line does not contain a `account_id`.

Is there ever a case where `if(!balance) {}` is triggered?  Do you ever use the `account_id`?

jniles commented on server/controllers/finance/reports/cashflow/index.js:
```diff
@@ -99,51 +103,33 @@ function queryIncomeExpense(params, dateFrom, dateTo) {
+        gl.account_id, gl.entity_uuid, gl.currency_id, gl.trans_id,
+        gl.description, gl.comment, gl.origin_id, gl.user_id
+      FROM general_ledger gl
+      WHERE gl.account_id IN (?) AND
```
It would be more efficient to say `account_id = ?`.  It doesn't look like you use multiple accounts.

jniles commented on server/controllers/finance/reports/cashflow/index.js:
```diff
@@ -223,15 +216,21 @@ function processingWeekCashflow(params) {
+    // get all periods for the the current fiscal year
+    return queryIncomeExpense(params);
+  })
+  .then((result) => {
```
Just FYI, the following two are equivalent:
```js
.then((result) => {
   return groupByPeriod(glb.periods, result, params.weekly);
});

// ... is the same as ...

.then(result =>  groupByPeriod(glb.periods, result, params.weekly));
```

jniles commented on server/controllers/finance/reports/cashflow/index.js:
```diff
@@ -553,11 +551,16 @@ function document(req, res, next) {
+        // if (tempIncome[item.origin_id] === true) {  >> Not tested TRUE
 
-        if (tempIncome[item.origin_id] === true) {
+        if (typeof tempIncome[item.origin_id] !== 'undefined') {
```
Why don't you store this in a temporary variable?

```js
var hasIncomeValue= typeof tempIncome[item.origin_id] !== 'undefined';
tempIncome[item.origin_id] = hasIncomeValue;
if (hasIncomeValue) {
  // do something
}
```

If the `undefined` check is giving you problems, you can try `_.isUndefined(tempIncome[item.origin_id])`.  [Here is the documentation](https://lodash.com/docs/4.17.4#isUndefined).

lomamech replied:
I just noticed that the data in the cashflow report differs from the account report, this difference is not obtained with the test data, I tried to add another transaction, a payment and a receipt with The main case in Dollars, I will go there to discover the origin of these errors

jniles reviewed (changes requested):
@lomamech This looks pretty good!

I've made a few comments to try and improve the code. From just the code, I think it could probably be merged after this.

lomamech committed:
Fix conflict in merging with branch master with translation fiile

lomamech committed:
test(Cash flow report)
- Reset data for the regression test in Cashflow Report

lomamech committed:
fix(Cash Flow Report)
- Updated the use of AccountsExtra.getOpeningBalanceForDate set property
  false in opening balance and true in closing balance

lomamech committed:
Resolve conflict in E2E shared components

lomamech committed:
Resolve conflict in en/Voucher.json

lomamech committed:
refactor(Cash Flow)
- Using .catch statement in bhCashBoxSelect
- Simplfy the using of then(result

jniles commented on client/src/modules/reports/generate/cashflow/cashflow.html:
```diff
@@ -0,0 +1,60 @@
+            </bh-date-interval>
+
+            <!-- by week  -->
+            <div class="checkbox">
```
It might be good to include some explanation text about what a weekly checkbox does.  You can use a `<p class="help-block"></p>` tag to add this text under the `<div class="checkbox></div>`.

For reference, [here is how the Open Debtors report](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/reports/generate/open_debtors/open_debtors.html#L89) accomplishes this.

jniles commented on test/temporary-financial-reports/cashFlow.js:
```diff
@@ -0,0 +1,171 @@
+    feb : 1061.45,
+    mar : 991.54,
+    apr : 924.43,
+    may : 855.9599999999999,
```
This is not a valid currency amount that we can display - if the server returns this, it means that we are doing some mathematics in JS.  We should make sure all the logic is in SQL.

I'm not sure if this should be blocking - @sfount, what do you think?

jniles reviewed (commented):
@lomamech, I made one comment about putting help text for users on the report and another about JS rounding bugs.

It looks like most of the Cashflow report is done in JS ... maybe we can schedule a refactor of this later.  It might be something I can work on in the US.  Don't worry about that comment until we figure out what the plan is for refactoring that report.  The other comment should be addressed.

lomamech committed:
Resolve Conflict

lomamech committed:
refactor(Cash Flow Report)
- Include help-class for explanation for some weekly checkbox
- Reformate the currency format in Regression Test

jniles reviewed (approved):
Looks good to me.  Good work :+1:
